### DISCUSSION TITLE
Interface with UCI code examples?

### DISCUSSION THREAD

**Discussion by o-k-a-y** [official-stockfish/Stockfish#3393] [Q&A]
Are there any code examples that deal with making calls to the UCI? Having a difficult time understanding the code base and would appreciate some guidance.

I'm trying to write a program that uses Stockfish to evaluate moves.

Someone told me you can interact with it through stdin and stdout but I learn best by example.

**Comment by vondele** [comment]
A simple example using bash commands:

```bash
printf "uci
        ucinewgame
        position fen r1bqkbnr/pppp1ppp/2n5/1B2p3/4P3/5N2/PPPP1PPP/RNBQK2R b KQkq - 0 1
        go depth 10
        ucinewgame
        quit" | ./stockfish | grep bestmove
```

**Reply by o-k-a-y** [reply]
What is ucinewgame doing here?

Are all the possible commands listed here?
http://page.mi.fu-berlin.de/block/uci.htm

And sorry for another question but how would it be possible to do something like:
1. set position e2e4 then analyze
2. set position c5c7 after e2e4 then anyalyze

Would I need to create the fen string each time I want to analyze a new move from the previous state? Is there some library that can convert from a fen string to another given move changes?

Thanks! The above example is helpful.

**Reply by maksimKorzh** [reply]
ucinewgame resets internal engine's settings like tratposition tables etc (this is in general, not sure what exactly happens in SF),
it's used in all known UCI engines.
GUI sends ucinewgame command to engine everytime you click "Start new game" button

re: another question

uci
ucinewgame
position startpos moves e2e4
go movetime 1000 // would be searching for 1 sec
go depth 15 // would search until reaches fixed depth of 15 plies

position startpos moves e2e4 c7c5
go movetime 1000 // would be searching for 1 sec
go depth 15 // would search until reaches fixed depth of 15 plies

etc.

you can also init position from fen string:
position fen r3k2r/p1ppqpb1/bn2pnp1/3PN3/1p2P3/2N2Q1p/PPPBBPPP/R3K2R w KQkq - 0 10
go movetime 1000

to print the board you can use stockfish specific command "d"

position startpos moves e2e4
d // this would print the current board state and fen of the position

**Reply by maksimKorzh** [reply]
@o-k-a-y 
I used chessboardjs and chessjs on the front end and python-chess on back-end.
The move list is kept by chessjs so the history is available as well as take backs.

Feel free to open an issue on the corresponding project's page for discussions
to avoid polluting sf's discussions with offtopic questions.

**Reply by o-k-a-y** [reply]
@vondele Is there somewhere I can see a complete list of commands or documentation I'm able to give the UCI?

Also correct me if I'm wrong, but that first ucinewgame command doesn't seem to be doing much at least with the example given. I ran the same command minus the first ucinewgame and had the same results.

Thanks again for all your help.

**Reply by vondele** [reply]
that first one is not needed for SF, but there might be other UCI engines where it is needed, it is required by the protocol.
